src/node/coinstats.cpp

```diff
@@ -37,45 +37,110 @@ CDataStream TxOutSer(const COutPoint& outpoint, const Coin& coin) {
     return ss;
 }
 
-//! Warning: be very careful when changing this! assumeutxo and UTXO snapshot
-//! validation commitments are reliant on the hash constructed by this
-//! function.
-//!
-//! If the construction of this hash is changed, it will invalidate
-//! existing UTXO snapshots. This will not result in any kind of consensus
-//! failure, but it will force clients that were expecting to make use of
-//! assumeutxo to do traditional IBD instead.
-//!
-//! It is also possible, though very unlikely, that a change in this
-//! construction could cause a previously invalid (and potentially malicious)
-//! UTXO snapshot to be considered valid.
-static void ApplyHash(CHashWriter& ss, const uint256& hash, const std::map<uint32_t, Coin>& outputs)
+class UTXOHasher
 {
-    for (auto it = outputs.begin(); it != outputs.end(); ++it) {
-        if (it == outputs.begin()) {
-            ss << hash;
-            ss << VARINT(it->second.nHeight * 2 + it->second.fCoinBase ? 1u : 0u);
-        }
+public:
+    virtual void Prepare(const uint256& hash_block){};
+    virtual void Apply(const uint256& hash, const std::map<uint32_t, Coin>& outputs){};
+    virtual uint256 Finalize() = 0;
+    virtual ~UTXOHasher();
+};
+
+UTXOHasher::~UTXOHasher() = default;
+
+class NullHasher final : public UTXOHasher {
+public:
+    uint256 Finalize() override {
+        return {};
+    }
+};
+
+std::unique_ptr<UTXOHasher> MakeNullHasher() {
+    return std::make_unique<NullHasher>();
+}
+
+class SHA256DHasher final : public UTXOHasher
```

Comment:
naming nit: I found the name `SHA256DHasher` a bit too specific and misleading, at first glance one could think that this is a helper for generally computing a SHA256D hash. IMHO the important criteria for naming is not what concrete hash function a UTXO hasher interally uses (e.g., MuHash also uses SHA256 for finalizing, and ChaCha20 for kind of "hashing" UTXOs to big integers via a stream cipher), but rather how it works on a more abstract conceptual level and in what way it commits to UTXOs etc. Not saying that's is easy to come up with a better name, but I'd prefer to keep `UTXO` in it, prefixed with a name resembling the output fields of the `gettxoutsetinfo` fields, e.g. `Legacy(Serialized)UTXOHasher` or `MuHashUTXOHasher`.

(Probably this comment will be obsolete anyway though if commit 9dc8093189e41849e91f8425f02e8f30563dfdf3 gets dropped)

Reply:
Yeah that's a good point, in any case I think it's no longer relevant as of 90358d4b7402bfd4cd9b79ebf828e40f4eabd849